### PR TITLE
added new mode to create NFTs on destination by caller

### PR DESCRIPTION
**Pull request by sasurobert**
added new mode to create NFTs on destination by caller

### CODE REVIEW

File: builtInFunctions/creator.go [hunk 1]
```diff
@@ -273,7 +273,7 @@ func (b *builtInFuncCreator) CreateBuiltInFunctionContainer() (vmcommon.BuiltInF
 		return nil, err
 	}
 
-	newFunc, err = NewESDTNFTCreateFunc(b.gasConfig.BuiltInCost.ESDTNFTCreate, b.gasConfig.BaseOperationCost, b.marshalizer, globalSettingsFunc, setRoleFunc, b.esdtStorageHandler, b.saveNFTToSystemAccountEnableEpoch, b.epochNotifier)
+	newFunc, err = NewESDTNFTCreateFunc(b.gasConfig.BuiltInCost.ESDTNFTCreate, b.gasConfig.BaseOperationCost, b.marshalizer, globalSettingsFunc, setRoleFunc, b.esdtStorageHandler, b.accounts, b.saveNFTToSystemAccountEnableEpoch, b.epochNotifier)
```
Comment (iulianpascalau) [inline builtInFunctions/creator.go:276]: this is a long list of params. We should switch to an argument type variable passing method.

File: builtInFunctions/esdtNFTCreate.go [hunk 2]
```diff
@@ -112,17 +119,42 @@ func (e *esdtNFTCreate) ProcessBuiltinFunction(
...
 		return nil, fmt.Errorf("%w, wrong number of arguments", ErrInvalidArguments)
 	}
 
+	accountWithRoles := acntSnd
+	uris := vmInput.Arguments[6:]
+	if vmInput.CallType == vm.ExecOnDestByCaller {
+		scAddressWithRoles := vmInput.Arguments[lenArgs-1]
+		uris = vmInput.Arguments[6 : lenArgs-1]
+
+		if len(scAddressWithRoles) != len(vmInput.CallerAddr) {
```
Comment (iulianpascalau) [inline builtInFunctions/esdtNFTCreate.go:138]: :+1:

File: builtInFunctions/esdtNFTCreate.go [hunk 3]
```diff
@@ -112,17 +119,42 @@ func (e *esdtNFTCreate) ProcessBuiltinFunction(
 	if err != nil {
 		return nil, err
 	}
-	if len(vmInput.Arguments) < 7 {
+
+	minNumOfArgs := 7
+	if vmInput.CallType == vm.ExecOnDestByCaller {
```
Comment (iulianpascalau) [inline builtInFunctions/esdtNFTCreate.go:124]: we need to condition the usage of the vm.ExecOnDestByCaller type against an epoch.

File: builtInFunctions/esdtNFTCreate.go [hunk 4]
```diff
@@ -112,17 +119,42 @@ func (e *esdtNFTCreate) ProcessBuiltinFunction(
...
+	uris := vmInput.Arguments[6:]
+	if vmInput.CallType == vm.ExecOnDestByCaller {
+		scAddressWithRoles := vmInput.Arguments[lenArgs-1]
+		uris = vmInput.Arguments[6 : lenArgs-1]
+
+		if len(scAddressWithRoles) != len(vmInput.CallerAddr) {
+			return nil, ErrInvalidAddressLength
+		}
+		if bytes.Equal(scAddressWithRoles, vmInput.CallerAddr) {
+			return nil, ErrInvalidRcvAddr
```
Comment (iulianpascalau) [inline builtInFunctions/esdtNFTCreate.go:142]: isn't this the sender? Maybe: `return ErrInvalidSndAddr` ?

**Review by iulianpascalau — COMMENTED** [review]
minor stuff

File: builtInFunctions/esdtNFTCreate.go (hunk 3 above)
Comment (sasurobert) [reply builtInFunctions/esdtNFTCreate.go:124]: it is conditioned in Arwen.

File: builtInFunctions/esdtNFTCreate.go (hunk 4 above)
Comment (sasurobert) [reply builtInFunctions/esdtNFTCreate.go:142]: receiver and sender is actually the same. so we can have both.

File: builtInFunctions/esdtNFTCreate.go [hunk 5]
```diff
@@ -112,17 +119,42 @@ func (e *esdtNFTCreate) ProcessBuiltinFunction(
...
+	accountWithRoles := acntSnd
+	uris := vmInput.Arguments[6:]
+	if vmInput.CallType == vm.ExecOnDestByCaller {
+		scAddressWithRoles := vmInput.Arguments[lenArgs-1]
+		uris = vmInput.Arguments[6 : lenArgs-1]
+
+		if len(scAddressWithRoles) != len(vmInput.CallerAddr) {
+			return nil, ErrInvalidAddressLength
+		}
+		if bytes.Equal(scAddressWithRoles, vmInput.CallerAddr) {
```
Comment (SebastianMarian) [inline builtInFunctions/esdtNFTCreate.go:141]: !byte.Equal

File: builtInFunctions/esdtNFTCreate_test.go [hunk 6]
```diff
@@ -294,6 +305,82 @@ func TestEsdtNFTCreate_ProcessBuiltinFunctionShouldWork(t *testing.T) {
...
+		&mock.EpochNotifierStub{},
+	)
+	address := bytes.Repeat([]byte{1}, 32)
+	userAddress := bytes.Repeat([]byte{2}, 32)
+	token := "token"
+	quantity := big.NewInt(2)
+	name := "name"
+	royalties := 100 //1%
+	hash := []byte("12345678901234567890123456789012")
+	attibutes := []byte("attributes")
```
Comment (SebastianMarian) [inline builtInFunctions/esdtNFTCreate_test.go:332]: attributes

File: builtInFunctions/esdtNFTCreate.go (hunk 5 above)
Comment (sasurobert) [reply builtInFunctions/esdtNFTCreate.go:141]: the current condition is correct. so you should not put the same address.

File: builtInFunctions/esdtNFTCreate_test.go (hunk 6 above)
Comment (sasurobert) [reply builtInFunctions/esdtNFTCreate_test.go]: done

File: builtInFunctions/creator.go (hunk 1 above)
Comment (sasurobert) [reply builtInFunctions/creator.go:276]: will do a refactor on next one.

File: builtInFunctions/esdtNFTCreate.go (hunk 5 above)
Comment (iulianpascalau) [reply builtInFunctions/esdtNFTCreate.go:141]: discussed, the vmInput.CallType == vm.ExecOnDestByCaller comes directly from the VM. So the VM is responsible for setting the last argument to the correct value. The protocol trusts the VM

**Review by iulianpascalau — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**